File: HWs/HW2/2.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>
#include <climits> // For INT_MAX

using namespace std;

// Function to read k-mers and queries
void readKmersAndQueries(int &kLength, unordered_map<string, int> &kmerFrequencies, vector<string> &queries) {
    int numKmers, numQueries;
    
    // Read number of k-mers and their length
    cin >> numKmers >> kLength;
    
    // Read k-mers and count their occurrences
    for (int i = 0; i < numKmers; ++i) {
        string kmer;
        cin >> kmer;
        kmerFrequencies[kmer]++;
    }
    
    // Read number of queries
    cin >> numQueries;
    
    // Read query strings
    for (int i = 0; i < numQueries; ++i) {
        string query;
        cin >> query;
        queries.push_back(query);
    }
}

// Function to calculate the maximum repetitions for a single query
int computeMaxRepeats(const string &query, int kLength, const unordered_map<string, int> &kmerFrequencies) {
    int querySize = query.size();
    
    // If the query is shorter than k, return -1
    if (querySize < kLength) {
        return -1;
    }
    
    // Extract k-mers from the query and count their occurrences
    unordered_map<string, int> requiredKmerCounts;
    for (int i = 0; i <= querySize - kLength; ++i) {
        string subKmer = query.substr(i, kLength);
        requiredKmerCounts[subKmer]++;
    }
    
    // Calculate the maximum repetitions
    int maxPossibleRepeats = INT_MAX;
    for (const auto &entry : requiredKmerCounts) {
        const string &subKmer = entry.first;
        int neededCount = entry.second;
        
        // If k-mer is not available, the query cannot be repeated
        if (kmerFrequencies.find(subKmer) == kmerFrequencies.end()) {
            return -1;
        }
        
        // Calculate the maximum repetitions for this k-mer
        maxPossibleRepeats = min(maxPossibleRepeats, kmerFrequencies.at(subKmer) / neededCount);
    }
    
    return maxPossibleRepeats;
}

// Function to process all queries and compute their results
vector<int> processQueries(int kLength, const unordered_map<string, int> &kmerFrequencies, const vector<string> &queries) {
    vector<int> results;
    for (const string &query : queries) {
        results.push_back(computeMaxRepeats(query, kLength, kmerFrequencies));
    }
    return results;
}

// Main function
int main() {
    int kLength;
    unordered_map<string, int> kmerFrequencies;
    vector<string> queries;
    
    // Read inputs
    readKmersAndQueries(kLength, kmerFrequencies, queries);
    
    // Process queries
    vector<int> results = processQueries(kLength, kmerFrequencies, queries);
    
    // Print results
    for (int result : results) {
        cout << result << endl;
    }
    
    return 0;
}

// g++ -o 2 2.cpp
// ./2

File: HWs/HW2/2.py
from collections import Counter, defaultdict

def read_kmers_and_queries():
    """
    Reads the number of k-mers, their length, and the k-mers themselves.
    Also reads the number of queries and the query strings.
    Returns k-mer data and query data as separate outputs.
    """
    # Read n and k values
    num_kmers, k_length = map(int, input().split())
    
    # Read k-mers and count their occurrences
    kmers = [input().strip() for _ in range(num_kmers)]
    kmer_frequencies = Counter(kmers)
    
    # Read number of queries
    num_queries = int(input())
    query_strings = [input().strip() for _ in range(num_queries)]
    
    return k_length, kmer_frequencies, query_strings

def compute_max_repeats(query, k_length, kmer_frequencies):
    """
    Calculates the maximum number of times a query string can repeat,
    given the available k-mers and their counts.
    """
    query_size = len(query)
    
    # If the query is shorter than k, return -1
    if query_size < k_length:
        return -1
    
    # Extract k-mers from the query
    query_sub_kmers = [query[i:i + k_length] for i in range(query_size - k_length + 1)]
    required_kmer_counts = Counter(query_sub_kmers)
    
    # Initialize maximum repetitions and a flag for unrestricted repeats
    max_possible_repeats = float('inf')
    unlimited_repeats = False
    
    # Check each k-mer in the query against available k-mers
    for substring, needed_count in required_kmer_counts.items():
        if substring in kmer_frequencies:
            max_possible_repeats = min(max_possible_repeats, kmer_frequencies[substring] // needed_count)
        else:
            unlimited_repeats = True
            break
    
    # If any k-mer is missing, return -1
    if unlimited_repeats:
        return -1
    
    return max_possible_repeats

def process_queries(k_length, kmer_frequencies, query_strings):
    """
    Processes each query to compute the maximum number of repetitions.
    """
    results = []
    for current_query in query_strings:
        results.append(compute_max_repeats(current_query, k_length, kmer_frequencies))
    return results

def main():
    """
    Main function to handle input, processing, and output.
    """
    # Read inputs
    k_length, kmer_frequencies, query_strings = read_kmers_and_queries()
    
    # Process the queries
    results = process_queries(k_length, kmer_frequencies, query_strings)
    
    # Print the results
    for res in results:
        print(res)

# Run the main function when executed
if __name__ == "__main__":
    main()
